src/reconcile/rec_visibility.c

```diff
@@ -93,36 +95,38 @@ __rec_append_orig_value(
      */
     append = tombstone = NULL; /* -Wconditional-uninitialized */
     total_size = size = 0;     /* -Wconditional-uninitialized */
-    if (unpack == NULL || unpack->type == WT_CELL_DEL)
-        WT_RET(__wt_update_alloc(session, NULL, &append, &size, WT_UPDATE_TOMBSTONE));
-    else {
-        WT_RET(__wt_scr_alloc(session, 0, &tmp));
-        WT_ERR(__wt_page_cell_data_ref(session, page, unpack, tmp));
-        WT_ERR(__wt_update_alloc(session, tmp, &append, &size, WT_UPDATE_STANDARD));
-        append->start_ts = append->durable_ts = unpack->start_ts;
-        append->txnid = unpack->start_txn;
-        total_size = size;
 
-        /*
-         * We need to append a TOMBSTONE before the onpage value if the onpage value has a valid
-         * stop pair.
-         *
-         * Imagine a case we insert and delete a value respectively at timestamp 0 and 10, and later
-         * insert it again at 20. We need the TOMBSTONE to tell us there is no value between 10 and
-         * 20.
-         */
-        if (unpack->stop_ts != WT_TS_MAX || unpack->stop_txn != WT_TXN_MAX) {
-            WT_ERR(__wt_update_alloc(session, NULL, &tombstone, &size, WT_UPDATE_TOMBSTONE));
-            tombstone->txnid = unpack->stop_txn;
-            tombstone->start_ts = unpack->stop_ts;
-            tombstone->durable_ts = unpack->stop_ts;
-            tombstone->next = append;
-            total_size += size;
-        }
+    /*
+     * We need to append a TOMBSTONE before the onpage value if the onpage value has a valid
+     * stop pair.
+     *
+     * Imagine a case we insert and delete a value respectively at timestamp 0 and 10, and later
+     * insert it again at 20. We need the TOMBSTONE to tell us there is no value between 10 and
+     * 20.
+     */
+    if (unpack->stop_ts != WT_TS_MAX || unpack->stop_txn != WT_TXN_MAX) {
+        /* No need to append anything if the stop time pair is globally visible. */
+        if (__wt_txn_visible_all(session, unpack->stop_txn, unpack->stop_ts))
+            return (0);
+        WT_ERR(__wt_update_alloc(session, NULL, &tombstone, &size, WT_UPDATE_TOMBSTONE));
+        tombstone->txnid = unpack->stop_txn;
+        tombstone->start_ts = unpack->stop_ts;
+        tombstone->durable_ts = unpack->durable_stop_ts;
+        total_size = size;
```

Comment:
we need to increment total_size with size, instead of assignment.

Reply:
fixed